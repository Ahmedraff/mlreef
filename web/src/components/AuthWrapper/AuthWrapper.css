.auth-wrapper {
  position: relative;
  opacity: 0.7;
  height: fit-content;
}

.auth-wrapper-cover {
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  background-color: #000;
  opacity: 0;
  cursor: auto;
}

.auth-wrapper-wrapped > * {
  display: block;
  margin: 0 !important;
}
