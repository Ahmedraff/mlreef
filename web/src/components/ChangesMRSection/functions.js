import { Base64 } from 'js-base64';
import Base64ToArrayBuffer from 'base64-arraybuffer';
import { getFileDifferences } from '../../functions/apiCalls';
import CommitsApi from 'apis/CommitsApi';
import { isImageFormat } from 'functions/dataParserHelpers';

const commitsApi = new CommitsApi();

export const fetchFileDiffInfo = (projectId, aheadCommits) => Promise.all(aheadCommits.map(async (commit, index) => {
  const { id: commitId } = commit;
  const commitDiffs = await commitsApi.getCommitDiff(projectId, commitId, 0, false);
  return await Promise.all(commitDiffs?.body?.map(async (commitDiff) => {
    const {
      previousVersionFile,
      nextVersionFile,
      imageFileSize,
    } = await getFileDifferences(projectId, commitDiff, commit.parent_ids[0], commit.id);
    let previousVersionFileParsed = previousVersionFile;
    let nextVersionFileParsed = nextVersionFile;
    let sizeDeleted = 0;
    let sizeAdded = 0;
    const fileName = commitDiff.old_path.split('/').slice(-1)[0];
    if (previousVersionFile && !isImageFormat(fileName)) {
      previousVersionFileParsed = Base64.decode(
        Base64ToArrayBuffer.encode(previousVersionFile),
      );
      sizeDeleted = previousVersionFile.byteLength;
    }
    if (nextVersionFile && !isImageFormat(fileName)) {
      nextVersionFileParsed = Base64.decode(Base64ToArrayBuffer.encode(nextVersionFile));
      sizeAdded = nextVersionFile.byteLength;
    }
    const path = commitDiff.new_path
      ? commitDiff.new_path
      : commitDiff.old_path;
    const fileExtension = path.split('.').pop().toLowerCase();
    return {
      fileName,
      fileExtension,
      previousVersionFileParsed,
      nextVersionFileParsed,
      id: `${commit.id} ${index}`,
      sizeDeleted,
      sizeAdded,
      imageFileSize,
    };
  }));
}));
