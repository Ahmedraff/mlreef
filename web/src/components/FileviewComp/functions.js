import FilesApi from 'apis/FilesApi.ts';
import CommitsApi from 'apis/CommitsApi.ts';

const commitsApi = new CommitsApi();

const filesApi = new FilesApi();

const getCommit = (projectId, lastCommitId) => commitsApi
  .getCommitDetails(projectId, lastCommitId);

const getFileAndInformation = (gid, file, branch, commit) => filesApi
  .getFileData(gid, file?.includes('/') ? encodeURIComponent(file) : file, branch || commit)
  .then(async (fData) => {
    const commitInfoDet = await getCommit(gid, fData.last_commit_id);
    return {
      fData,
      commitInfoDet,
    };
  });

export default {
  getFileAndInformation,
};
