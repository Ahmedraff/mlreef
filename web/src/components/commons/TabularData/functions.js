import genUUID from 'uuid/v1';

const parseNumber = (text) => {
  const num = parseFloat(text);

  return Number.isNaN(num) ? text : num;
};

export const transpose = (matrix) => matrix[0].map((_, ci) => matrix.map((row) => row[ci]));

const sanatizeMatrix = (matrix) => {
  const head = matrix[0];
  const tail = matrix[matrix.length - 1];

  const sameLength = head.length === tail.length;
  const lastRowfilled = tail.every((col) => col !== undefined);

  return sameLength && lastRowfilled ? matrix : matrix.slice(0, -1);
};

export const cvsToArray = (text, separator = ',') => {
  const matrix = text.split('\n')
    .map((line) => line.split(separator)
      .map(parseNumber));

  return sanatizeMatrix(matrix);
};

export const jsonToArray = (input) => {
  const items = Array.isArray(input) ? input : Object.values(input);

  if (!items.length) return items;
  // const row = Object.values(input)[0];
  const row = items[0];
  const header = Object.keys(row).map((label) => label);

  // const data = Object.values(input)
  const data = items
    .map((cols) => Object.values(cols).map((val) => val));

  const matrix = [header].concat(data);

  return sanatizeMatrix(matrix);
};

export const arrayToRichData = (matrix, separator = ',') => {
  const header = matrix[0];

  return {
    id: genUUID(),
    meta: {
      separator,
      timestamp: new Date(),
    },
    matrix,
    // data: flat(matrix.map((row, y) => row.map((value, x) => ({ x, y, value }))))
    data: matrix.map((cols, y) => ({
      id: y,
      cols: cols.map((value, x) => ({
        x,
        y,
        value,
        label: header[x],
      })),
    })),
  };
};
