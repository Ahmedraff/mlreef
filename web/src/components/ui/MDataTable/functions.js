const extractField = (row, x) => row.cols.find((col) => col.x === x)?.value;

export const sortDesc = (x) => (rows) => rows.slice()
  .sort((a, b) => extractField(b, x) - extractField(a, x));

export const sortAsc = (x) => (rows) => rows.slice()
  .sort((a, b) => extractField(a, x) - extractField(b, x));

export const round = (decs) => (input) => {
  const decimals = parseInt(decs, 10);

  if (Number.isNaN(decimals)) return input;

  if (decimals > 12) return input;

  const num = parseFloat(input);
  if (Number.isNaN(num)) return input;

  return Math.round(num * (10 ** decimals)) / (10 ** decimals);
};
