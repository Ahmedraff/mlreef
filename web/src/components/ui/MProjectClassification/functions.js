import { ALGORITHM, DATA_OPERATION, MODEL, OPERATION, VISUALIZATION } from "dataTypes";

export const getSearchable = (feSearchableTypes) => {
  switch (feSearchableTypes) {
    case DATA_OPERATION:
      return OPERATION;
    case MODEL:
      return ALGORITHM;
    default:
      return VISUALIZATION;
  }
}