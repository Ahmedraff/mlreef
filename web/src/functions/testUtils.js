import { createStore, applyMiddleware } from 'redux';

import rootReducer from '../reducers/index';
import { middlewares } from '../configureStore';

// eslint-disable-next-line import/prefer-default-export
export const storeFactory = (initialState) => {
  const createStoreWithMiddleware = applyMiddleware(...middlewares)(createStore);
  return createStoreWithMiddleware(rootReducer, initialState);
};
