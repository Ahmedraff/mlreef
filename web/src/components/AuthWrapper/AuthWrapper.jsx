import React, { useMemo } from 'react';
import { useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';
import propTypes from 'prop-types';
import './AuthWrapper.css';
import { fireModal } from 'actions/actionModalActions';

const isOwner = false;
const hardcodedRole = 2;
const hardcodedAccountType = 1;

const AuthWrapper = (props) => {
  const {
    style,
    owneronly,
    role,
    accountType,
    norender,
    children,
  } = props;
  
  const history = useHistory();
  const dispatch = useDispatch();
  
  const owned = useMemo(() => owneronly ? isOwner : true, [owneronly]);
  
  const hasRole = useMemo(() => hardcodedRole >= role, [role]);

  const hasAccountType = useMemo(() => hardcodedAccountType >= accountType, [accountType]);

  const allowed = useMemo(
    () => owned && hasRole && hasAccountType,
    [owned, hasRole, hasAccountType],
  );

  const classes = useMemo(() => ({
    main: `auth-wrapper
      ${!owned ? 'ownership-required' : ''}
      ${!hasRole ? 'group-role-required' : ''}
      ${!hasAccountType ? 'project-role-required' : ''}
    `,
  }), [owned, hasRole, hasAccountType]);

  const message = useMemo(
    () => {
      if (!owned) return 'Only the owner, you can fork it!';
      if (!hasRole) return 'You need a proper role.';
      if (!hasAccountType) return 'Upgrade your account';

      return 'You need permission to use this feature';
    },
    [owned, hasRole, hasAccountType],
  );
  
  const handleClick = (e) => {    
    if (!hasAccountType) return dispatch(fireModal({
      type: 'info',
      title: 'Upgraded needed',
      dark: false,
      content: (
        <div className="m-auto t-center">
          <h3>Please upgrade your account!</h3>
        </div>
      ),
      positiveLabel: 'UPGRADE',
      negativeLabel: 'BACK',
      onPositive: () => history.push('/account/upgrade'),    
    }));
  };

  return allowed ? children : (norender ? null : (
    <div title={message} className={classes.main} style={style}>
      <div onClick={handleClick} className="auth-wrapper-cover"></div>
      <div className="auth-wrapper-wrapped">
        {children}
      </div>
    </div>
  ));
};

AuthWrapper.defaultProps = {
  owner: false,
  role: 0,
  accountType: 0,
  norender: false,
  style: {},
}

AuthWrapper.propTypes = {
  owner: propTypes.bool,
  role: propTypes.number,
  accountType: propTypes.number,
  norender: propTypes.bool,
  // children:
}

export default AuthWrapper;
