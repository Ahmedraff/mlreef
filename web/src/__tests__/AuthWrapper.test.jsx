import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { storeFactory } from 'functions/testUtils';
import AuthWrapper from 'components/AuthWrapper';
import initialState from 'store/reducers/initialState';

/*
 * These snapshots should never break!
 *
 * What to do if snapshots changed?
 *
 * Do not update it without a prelimitar check!
 *
 * What could had happened to break a snapshot?
 *
 * Something could have changed in AuthWrapper, permission hooks or gitlab
 * object in selectedProject, therefore the best is to fix those relationships.
 */

// eslint-disable-next-line
const Subject = ({ desc }) => (
  <div>{desc || 'no description'}</div>
);

// a base state
const state = {
  ...initialState,
  projects: {
    ...initialState.projects,
    selectedProject: {
      id: 'project-uuid-001',
      gitlab: {
        owner: {
          username: 'mlreef',
        },
      },
      namespace: {
        path: 'mlreef',
      },
    },
  },
};

const stores = {
  visitor: storeFactory(state),

  authenticated: storeFactory({
    ...state,
    user: {
      ...state.user,
      username: 'calca',
      auth: true,
      token: '123',
    },
  }),

  owner: storeFactory({
    projects: {
      selectedProject: {
        ...state.projects.selectedProject,
        gitlab: {
          ...state.projects.selectedProject.gitlab,
          permissions: {
            project_access: {
              access_level: 40,
            },
          },
        },
      },
    },
    user: {
      ...state.user,
      username: 'mlreef',
      auth: true,
      token: '123',
    },
  }),

  developer: storeFactory({
    ...state,
    projects: {
      ...state.projects,
      selectedProject: {
        ...state.projects.selectedProject,
        gitlab: {
          ...state.projects.selectedProject.gitlab,
          permissions: {
            project_access: {
              access_level: 30,
            },
          },
        },
      },
    },
    user: {
      ...state.user,
      username: 'imadev',
      auth: true,
      token: 'aae1',
    },
  }),
};

describe('Evaluate AuthWrapper', () => {
  test('assert snapshot for visitor', () => {
    const component = renderer.create(
      <main>
        <Provider store={stores.visitor}>
          <section>
            visitor: only auth
            <AuthWrapper>
              <Subject desc="should render covered" />
            </AuthWrapper>
          </section>
          <section>
            visitor: only auth, norender
            <AuthWrapper norender>
              <Subject desc="should not render" />
            </AuthWrapper>
          </section>
          <section>
            visitor: developer covered
            <AuthWrapper minRole={30}>
              <Subject desc="should render covered" />
            </AuthWrapper>
          </section>
          <section>
            visitor: developer, norender
            <AuthWrapper minRole={30} norender>
              <Subject desc="should not render" />
            </AuthWrapper>
          </section>
          <section>
            visitor: owneronly, covered
            <AuthWrapper owneronly>
              <Subject desc="should render covered" />
            </AuthWrapper>
          </section>
          <section>
            visitor: owneronly norender
            <AuthWrapper owneronly norender>
              <Subject desc="should not render" />
            </AuthWrapper>
          </section>
        </Provider>
      </main>,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('assert snapshot for authenticated', () => {
    const component = renderer.create(
      <main>
        <Provider store={stores.authenticated}>
          <section>
            authenticated: only auth
            <AuthWrapper>
              <Subject desc="should render" />
            </AuthWrapper>
          </section>
          <section>
            authenticated: only auth, norender
            <AuthWrapper norender>
              <Subject desc="should render" />
            </AuthWrapper>
          </section>
          <section>
            authenticated: developer
            <AuthWrapper minRole={30}>
              <Subject desc="should render covered" />
            </AuthWrapper>
          </section>
          <section>
            authenticated: developer, norender
            <AuthWrapper minRole={30} norender>
              <Subject desc="should not render" />
            </AuthWrapper>
          </section>
          <section>
            authenticated: owneronly norender
            <AuthWrapper owneronly>
              <Subject desc="should render covered" />
            </AuthWrapper>
          </section>
          <section>
            authenticated: owneronly norender
            <AuthWrapper owneronly norender>
              <Subject desc="should not render" />
            </AuthWrapper>
          </section>
        </Provider>
      </main>,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('assert snapshot for member of a project', () => {
    const component = renderer.create(
      <main>
        <Provider store={stores.developer}>
          <section>
            developer: only auth
            <AuthWrapper>
              <Subject desc="should render" />
            </AuthWrapper>
          </section>
          <section>
            developer: only auth, norender
            <AuthWrapper norender>
              <Subject desc="should render" />
            </AuthWrapper>
          </section>
          <section>
            developer: developer
            <AuthWrapper minRole={30}>
              <Subject desc="should render" />
            </AuthWrapper>
          </section>
          <section>
            developer: developer, covered
            <AuthWrapper minRole={30} norender>
              <Subject desc="should render" />
            </AuthWrapper>
          </section>
          <section>
            developer: developer w/o perms, covered
            <AuthWrapper minRole={40}>
              <Subject desc="should render covered" />
            </AuthWrapper>
          </section>
          <section>
            developer: developer w/o perms, no render
            <AuthWrapper minRole={40} norender>
              <Subject desc="should not render" />
            </AuthWrapper>
          </section>
          <section>
            developer: owneronly covered
            <AuthWrapper owneronly>
              <Subject desc="should render covered" />
            </AuthWrapper>
          </section>
          <section>
            developer: owneronly norender
            <AuthWrapper owneronly norender>
              <Subject desc="should not render" />
            </AuthWrapper>
          </section>
        </Provider>
      </main>,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('assert snapshot for project owner', () => {
    const component = renderer.create(
      <main>
        <Provider store={stores.owner}>
          <section>
            owner: only auth
            <AuthWrapper>
              <Subject desc="should render" />
            </AuthWrapper>
          </section>
          <section>
            owner: only auth, norender
            <AuthWrapper norender>
              <Subject desc="should render" />
            </AuthWrapper>
          </section>
          <section>
            owner: developer
            <AuthWrapper minRole={30}>
              <Subject desc="should render" />
            </AuthWrapper>
          </section>
          <section>
            owner: developer, covered
            <AuthWrapper minRole={30} norender>
              <Subject desc="should render" />
            </AuthWrapper>
          </section>
          <section>
            owner: owneronly covered
            <AuthWrapper owneronly>
              <Subject desc="should render" />
            </AuthWrapper>
          </section>
          <section>
            owner: owneronly norender
            <AuthWrapper owneronly norender>
              <Subject desc="should render" />
            </AuthWrapper>
          </section>
        </Provider>
      </main>,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
