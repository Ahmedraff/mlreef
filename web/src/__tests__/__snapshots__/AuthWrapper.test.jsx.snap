// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Evaluate AuthWrapper assert snapshot for authenticated 1`] = `
<main>
  <section>
    authenticated: only auth
    <div>
      should render
    </div>
  </section>
  <section>
    authenticated: only auth, norender
    <div>
      should render
    </div>
  </section>
  <section>
    authenticated: developer
    <div
      className="auth-wrapper ownership-required group-role-required"
      title="Only the owner, you can fork it!"
    >
      <div
        className="auth-wrapper-cover"
        onClick={[Function]}
      />
      <div
        className="auth-wrapper-wrapped"
      >
        <div>
          should render covered
        </div>
      </div>
    </div>
  </section>
  <section>
    authenticated: developer, norender
  </section>
  <section>
    authenticated: owneronly norender
    <div
      className="auth-wrapper ownership-required"
      title="Only the owner, you can fork it!"
    >
      <div
        className="auth-wrapper-cover"
        onClick={[Function]}
      />
      <div
        className="auth-wrapper-wrapped"
      >
        <div>
          should render covered
        </div>
      </div>
    </div>
  </section>
  <section>
    authenticated: owneronly norender
  </section>
</main>
`;

exports[`Evaluate AuthWrapper assert snapshot for member of a project 1`] = `
<main>
  <section>
    developer: only auth
    <div>
      should render
    </div>
  </section>
  <section>
    developer: only auth, norender
    <div>
      should render
    </div>
  </section>
  <section>
    developer: developer
    <div>
      should render
    </div>
  </section>
  <section>
    developer: developer, covered
    <div>
      should render
    </div>
  </section>
  <section>
    developer: developer w/o perms, covered
    <div
      className="auth-wrapper ownership-required group-role-required"
      title="Only the owner, you can fork it!"
    >
      <div
        className="auth-wrapper-cover"
        onClick={[Function]}
      />
      <div
        className="auth-wrapper-wrapped"
      >
        <div>
          should render covered
        </div>
      </div>
    </div>
  </section>
  <section>
    developer: developer w/o perms, no render
  </section>
  <section>
    developer: owneronly covered
    <div
      className="auth-wrapper ownership-required"
      title="Only the owner, you can fork it!"
    >
      <div
        className="auth-wrapper-cover"
        onClick={[Function]}
      />
      <div
        className="auth-wrapper-wrapped"
      >
        <div>
          should render covered
        </div>
      </div>
    </div>
  </section>
  <section>
    developer: owneronly norender
  </section>
</main>
`;

exports[`Evaluate AuthWrapper assert snapshot for project owner 1`] = `
<main>
  <section>
    owner: only auth
    <div>
      should render
    </div>
  </section>
  <section>
    owner: only auth, norender
    <div>
      should render
    </div>
  </section>
  <section>
    owner: developer
    <div>
      should render
    </div>
  </section>
  <section>
    owner: developer, covered
    <div>
      should render
    </div>
  </section>
  <section>
    owner: owneronly covered
    <div>
      should render
    </div>
  </section>
  <section>
    owner: owneronly norender
    <div>
      should render
    </div>
  </section>
</main>
`;

exports[`Evaluate AuthWrapper assert snapshot for visitor 1`] = `
<main>
  <section>
    visitor: only auth
    <div
      className="auth-wrapper ownership-required"
      title="Please login"
    >
      <div
        className="auth-wrapper-cover"
        onClick={[Function]}
      />
      <div
        className="auth-wrapper-wrapped"
      >
        <div>
          should render covered
        </div>
      </div>
    </div>
  </section>
  <section>
    visitor: only auth, norender
  </section>
  <section>
    visitor: developer covered
    <div
      className="auth-wrapper ownership-required group-role-required"
      title="Please login"
    >
      <div
        className="auth-wrapper-cover"
        onClick={[Function]}
      />
      <div
        className="auth-wrapper-wrapped"
      >
        <div>
          should render covered
        </div>
      </div>
    </div>
  </section>
  <section>
    visitor: developer, norender
  </section>
  <section>
    visitor: owneronly, covered
    <div
      className="auth-wrapper ownership-required"
      title="Please login"
    >
      <div
        className="auth-wrapper-cover"
        onClick={[Function]}
      />
      <div
        className="auth-wrapper-wrapped"
      >
        <div>
          should render covered
        </div>
      </div>
    </div>
  </section>
  <section>
    visitor: owneronly norender
  </section>
</main>
`;
